File: server/controllers/group.go
package controllers

import (
	"onxzy/super-santa-server/controllers/dto"
	"onxzy/super-santa-server/database/models"
	"onxzy/super-santa-server/middlewares"
	"onxzy/super-santa-server/services"
	"onxzy/super-santa-server/services/authService"
	"onxzy/super-santa-server/services/groupService"
	"onxzy/super-santa-server/services/userService"

	"github.com/gin-gonic/gin"
)

type GroupController struct {
	authService  *services.AuthService
	groupService *services.GroupService
	userService  *services.UserService
}

func NewGroupController(groupService *services.GroupService, authService *services.AuthService, userService *services.UserService) *GroupController {
	return &GroupController{
		groupService: groupService,
		authService:  authService,
		userService:  userService,
	}
}
func (gc *GroupController) RegisterRoutes(router *gin.RouterGroup, authMiddleware *middlewares.AuthMiddleware) {
	router.POST("", gc.CreateGroup)
	router.GET("/info/:group_id", gc.GetGroupInfo)
	router.POST("/join", gc.JoinGroup)

	authRouter := router.Group("").Use(authMiddleware.Auth)
	authRouter.GET("", gc.GetGroup)
	authRouter.PUT("/wishes", gc.UpdateWishes)
	authRouter.GET("/draw", gc.InitDraw)
	authRouter.POST("/draw", gc.FinishDraw)
}

// Create Group

type CreateGroupResponse = models.Group

func (gc *GroupController) CreateGroup(c *gin.Context) {
	// Validate request
	var req dto.CreateGroupRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(400, gin.H{"error": err.Error()})
		return
	}

	group := &models.Group{
		Name:           req.Name,
		SecretVerifier: req.SecretVerifier,
		Results:        nil,
	}

	admin := &models.User{
		Username:            req.Admin.Username,
		Email:               req.Admin.Email,
		PasswordVerifier:    req.Admin.PasswordVerifier,
		PublicKeySecret:     req.Admin.PublicKeySecret,
		PrivateKeyEncrypted: req.Admin.PrivateKeyEncrypted,
		IsAdmin:             true,
	}

	if err := gc.groupService.CreateGroup(group, admin); err != nil {
		// TODO: Add error handling
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}

	// Return group
	c.JSON(201, group)
}

// Get Group Info

type GetGroupInfoResponse = groupService.GroupInfo

func (gc *GroupController) GetGroupInfo(c *gin.Context) {
	// Get group ID from URL
	groupID := c.Param("group_id")
	// Get group info from service
	groupInfo, err := gc.groupService.GetGroupInfo(groupID)
	if err != nil {
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}
	// Return group info
	c.JSON(200, groupInfo)
}

// Get Group

func (gc *GroupController) GetGroup(c *gin.Context) {
	claims := c.MustGet("claims").(*authService.AuthClaims)
	groupID := claims.GroupID

	// Get group from service
	group, err := gc.groupService.GetGroup(groupID)
	if err != nil {
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}

	c.JSON(200, group)
}

func (gc *GroupController) JoinGroup(c *gin.Context) {
	req := &dto.JoinGroupRequest{}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(400, gin.H{"error": err.Error()})
		return
	}

	groupID, err := gc.authService.VerifyGroupJWT(req.GroupToken)
	if err != nil {
		c.JSON(401, gin.H{"error": "Unauthorized"})
		return
	}

	// Check that group exists
	_, err = gc.groupService.GetGroup(groupID)
	if err != nil {
		if err == groupService.ErrGroupNotFound {
			c.JSON(404, gin.H{"error": "Group not found"})
			return
		}
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}

	user := &models.User{
		Username:            req.User.Username,
		Email:               req.User.Email,
		PasswordVerifier:    req.User.PasswordVerifier,
		PublicKeySecret:     req.User.PublicKeySecret,
		PrivateKeyEncrypted: req.User.PrivateKeyEncrypted,
		GroupID:             groupID,
		IsAdmin:             false,
	}

	err = gc.userService.CreateUser(user)
	if err != nil {
		if err == userService.ErrUserAlreadyExists {
			c.JSON(409, gin.H{"error": "User already exists"})
			return
		}
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}

	c.JSON(201, user)

}

func (gc *GroupController) UpdateWishes(c *gin.Context) {
	claims := c.MustGet("claims").(*authService.AuthClaims)
	userID := claims.Subject

	var req dto.UpdateWishesRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(400, gin.H{"error": err.Error()})
		return
	}

	// Get user from service
	user, err := gc.userService.GetUser(userID)
	if err != nil {
		if err == userService.ErrUserNotFound {
			c.JSON(404, gin.H{"error": "User not found"})
			return
		}
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}

	// Update user
	user.Wishes = req.Wishes
	if err := gc.userService.UpdateUser(user); err != nil {
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}

	c.JSON(200, &dto.UpdateWishesResponse{
		Wishes: user.Wishes,
	})
}

// TODO: Add error handling
func (gc *GroupController) InitDraw(c *gin.Context) {
	claims := c.MustGet("claims").(*authService.AuthClaims)
	groupID := claims.GroupID

	if !claims.IsAdmin {
		c.JSON(403, gin.H{"error": "Forbidden"})
		return
	}

	// Get group from service
	group, err := gc.groupService.GetGroup(groupID)
	if err != nil {
		if err == groupService.ErrGroupNotFound {
			c.JSON(404, gin.H{"error": "Group not found"})
			return
		}
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}

	if group.Results != nil {
		c.JSON(409, gin.H{"error": "Draw already done"})
		return
	}

	publicKeys, err := gc.groupService.InitDraw(groupID)
	if err != nil {
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}

	c.JSON(200, &dto.InitDrawResponse{
		PublicKeysSecret: publicKeys,
	})
}

// TODO: Add error handling
func (gc *GroupController) FinishDraw(c *gin.Context) {
	claims := c.MustGet("claims").(*authService.AuthClaims)
	groupID := claims.GroupID

	if !claims.IsAdmin {
		c.JSON(403, gin.H{"error": "Forbidden"})
		return
	}

	group, err := gc.groupService.GetGroup(groupID)
	if err != nil {
		if err == groupService.ErrGroupNotFound {
			c.JSON(404, gin.H{"error": "Group not found"})
			return
		}
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}

	if group.Results != nil {
		c.JSON(409, gin.H{"error": "Draw already done"})
		return
	}

	var req dto.FinishDrawRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(400, gin.H{"error": err.Error()})
		return
	}

	if _, err := gc.groupService.FinishDraw(groupID, req.PublicKeys); err != nil {
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}

	c.Status(200)
}

File: api-sdk/lib/api/group.ts
import { AuthAPIError, AuthAPIErrorCode } from "./auth";
import { ApiClient, ApiError } from "./client";
import {
  CreateGroupRequest,
  FinishDrawRequest,
  GroupAPIStatusCode,
  GroupInfo,
  GroupModel,
  InitDrawResponse,
  JoinGroupRequest,
} from "./dto/group";
import { UpdateWishesRequest, UpdateWishesResponse } from "./dto/user";

export enum GroupAPIErrorCode {
  GROUP_AUTH_ERROR = "GROUP_AUTH_ERROR",

  NOT_ENOUGH_USERS = "NOT_ENOUGH_USERS",
  DRAW_NOT_INITIED = "DRAW_NOT_INITIED",
  DRAW_DONE = "DRAW_DONE",

  UNKNOWN_ERROR = "UNKNOWN_ERROR",
}

export class GroupAPIError extends Error {
  constructor(
    public code: GroupAPIErrorCode,
    error: unknown = null,
    message: string = "Unknown error"
  ) {
    super(
      error instanceof Error
        ? `${message} : [${error.name}] ${error.message}`
        : message
    );
    this.name = "GroupAPIError";
  }
}

export class GroupAPI {
  private static basePath = "/group";
  private authContext;

  constructor(private client: ApiClient) {
    this.authContext = client.getAuthContext();
  }

  async createGroup(
    name: string,
    admin: { username: string; email: string },
    encodedKeys: {
      secretVerifier: string;
      passwordVerifier: string;
      privateKeyEncrypted: string;
      publicKeySecret: string;
    }
  ): Promise<GroupModel> {
    const group = await this.client.post<CreateGroupRequest, GroupModel>(
      `${GroupAPI.basePath}`,
      {
        name,
        secret_verifier: encodedKeys.secretVerifier,
        admin: {
          email: admin.email,
          username: admin.username,
          password_verifier: encodedKeys.passwordVerifier,
          public_key_secret: encodedKeys.publicKeySecret,
          private_key_encrypted: encodedKeys.privateKeyEncrypted,
        },
      }
    );

    return group;
  }

  async joinGroup(
    user: { username: string; email: string },
    encodedKeys: {
      passwordVerifier: string;
      privateKeyEncrypted: string;
      publicKeySecret: string;
    }
  ): Promise<GroupModel> {
    const groupToken = this.authContext.getGroupToken();
    if (!groupToken) {
      throw new GroupAPIError(
        GroupAPIErrorCode.GROUP_AUTH_ERROR,
        null,
        "Group token is not set. Please login to group first."
      );
    }

    const group = await this.client.post<JoinGroupRequest, GroupModel>(
      `${GroupAPI.basePath}/join`,
      {
        group_token: groupToken,
        user: {
          username: user.username,
          email: user.email,
          password_verifier: encodedKeys.passwordVerifier,
          public_key_secret: encodedKeys.publicKeySecret,
          private_key_encrypted: encodedKeys.privateKeyEncrypted,
        },
      }
    );

    return group;
  }

  async getGroupInfo(groupID: string): Promise<GroupInfo | null> {
    try {
      return await this.client.get<GroupInfo>(
        `${GroupAPI.basePath}/${groupID}/info`
      );
    } catch (error) {
      if (error instanceof ApiError) {
        if (error.status === 404) return null; // Group not found
      }
      throw new GroupAPIError(
        GroupAPIErrorCode.UNKNOWN_ERROR,
        error,
        "Failed to get group info"
      );
    }
  }

  async getGroup(): Promise<GroupModel> {
    try {
      return await this.client.get<GroupModel>(`${GroupAPI.basePath}`);
    } catch (error) {
      if (error instanceof ApiError) {
        // 400 should not occur
        if (error.status === 401)
          throw new AuthAPIError(AuthAPIErrorCode.AUTH_ERROR, error);
      }
      throw new GroupAPIError(
        GroupAPIErrorCode.UNKNOWN_ERROR,
        error,
        "Failed to get group"
      );
    }
  }

  async updateWishes(wishes: string): Promise<string> {
    try {
      const { wishes: newWishes } = await this.client.put<
        UpdateWishesRequest,
        UpdateWishesResponse
      >(`${GroupAPI.basePath}/wishes`, { wishes });
      return newWishes;
    } catch (error) {
      if (error instanceof ApiError) {
        // 400 and 404 should not occur
        if (error.status === 401)
          throw new AuthAPIError(AuthAPIErrorCode.AUTH_ERROR, error);
      }
      throw new GroupAPIError(
        GroupAPIErrorCode.UNKNOWN_ERROR,
        error,
        "Failed to update wishes"
      );
    }
  }

  /**
   *
   * @throws {GroupAPIError} NOT_ENOUGH_USERS, DRAW_NOT_INITIED
   */
  async initDraw(): Promise<string[]> {
    try {
      const { public_keys_secret } = await this.client.get<InitDrawResponse>(
        `${GroupAPI.basePath}/draw`
      );
      return public_keys_secret;
    } catch (error) {
      if (error instanceof ApiError) {
        if (error.status === GroupAPIStatusCode.NOT_ENOUGH_USERS)
          throw new GroupAPIError(
            GroupAPIErrorCode.NOT_ENOUGH_USERS,
            error,
            "Failed to init draw"
          );
        if (error.status === 409)
          throw new GroupAPIError(
            GroupAPIErrorCode.DRAW_DONE,
            error,
            "Draw already done"
          );
        if (error.status === 401)
          throw new AuthAPIError(AuthAPIErrorCode.AUTH_ERROR, error);
        if (error.status === 403)
          throw new AuthAPIError(AuthAPIErrorCode.FORBIDDEN, error);
      }
      throw new GroupAPIError(
        GroupAPIErrorCode.UNKNOWN_ERROR,
        error,
        "Failed to init draw"
      );
    }
  }

  /**
   *
   * @throws {GroupAPIError} DRAW_NOT_INITIED, DRAW_DONE
   */
  async finishDraw(publicKeys: JsonWebKey[]): Promise<void> {
    try {
      await this.client.post<FinishDrawRequest, null>(
        `${GroupAPI.basePath}/draw`,
        {
          public_keys: publicKeys.map((key) => JSON.stringify(key)),
        }
      );
    } catch (error) {
      if (error instanceof ApiError) {
        if (error.status === GroupAPIStatusCode.DRAW_SESSION_NOT_FOUND)
          throw new GroupAPIError(
            GroupAPIErrorCode.DRAW_NOT_INITIED,
            error,
            "Draw not initied or too old. init draw again"
          );
        if (error.status === 409)
          throw new GroupAPIError(
            GroupAPIErrorCode.DRAW_DONE,
            error,
            "Draw already done"
          );
        if (error.status === 401)
          throw new AuthAPIError(AuthAPIErrorCode.AUTH_ERROR, error);
        if (error.status === 403)
          throw new AuthAPIError(AuthAPIErrorCode.FORBIDDEN, error);
      }
      throw new GroupAPIError(
        GroupAPIErrorCode.UNKNOWN_ERROR,
        error,
        "Failed to finish draw"
      );
    }
  }

  /**
   *
   * @throws {AuthAPIError} AUTH_ERROR, FORBIDDEN
   * @throws {GroupAPIError} DRAW_DONE
   */
  async deleteUser(userID: string): Promise<void> {
    try {
      await this.client.delete(`${GroupAPI.basePath}/user/${userID}`);
    } catch (error) {
      if (error instanceof ApiError) {
        if (error.status === 401)
          throw new AuthAPIError(AuthAPIErrorCode.AUTH_ERROR, error);
        if (error.status === 403)
          throw new AuthAPIError(AuthAPIErrorCode.FORBIDDEN, error);
        if (error.status === 409)
          throw new GroupAPIError(
            GroupAPIErrorCode.DRAW_DONE,
            error,
            "Draw already done"
          );
      }
      throw new GroupAPIError(
        GroupAPIErrorCode.UNKNOWN_ERROR,
        error,
        "Failed to delete user"
      );
    }
  }

  /**
   *
   * @throws {AuthAPIError} AUTH_ERROR, FORBIDDEN
   * @throws {GroupAPIError} DRAW_DONE
   * */
  async leaveGroup(): Promise<void> {
    try {
      await this.client.delete(`${GroupAPI.basePath}/user`);
    } catch (error) {
      if (error instanceof ApiError) {
        if (error.status === 401)
          throw new AuthAPIError(AuthAPIErrorCode.AUTH_ERROR, error);
        if (error.status === 403)
          throw new AuthAPIError(AuthAPIErrorCode.FORBIDDEN, error);
        if (error.status === 409)
          throw new GroupAPIError(
            GroupAPIErrorCode.DRAW_DONE,
            error,
            "Draw already done"
          );
      }
      throw new GroupAPIError(
        GroupAPIErrorCode.UNKNOWN_ERROR,
        error,
        "Failed to leave group"
      );
    }
  }
}
